index.bs

```diff
@@ -2202,159 +2461,140 @@ interface MLGraph {
-                    1. Set the dimensions of |inputTensor| to |inputDesc|.{{MLOperandDescriptor/dimensions}}.
-                1. If |value| is an {{MLInput}}, then:
-                    1. Set the values of |inputTensor| to the values of |value|.{{MLInput/resource}}.
-                1. If |value| is an {{MLResource}}, then:
-                    1. Set the values of |inputTensor| to the values of |value|.
-                1. Set the input of |this|.{{MLGraph/[[implementation]]}} that is associated with |key| to |inputTensor|.
-            1. For each |key| -> |value| of |outputs|:
-                1. Issue a compute request for output of |this|.{{MLGraph/[[implementation]]}} that is associated with |key|.
-                1. Wait for the compute request to be completed.
-                1. If there is an error returned by |this|.{{MLGraph/[[implementation]]}}, then:
-                    1. Throw an {{OperationError}} {{DOMException}} and stop.
-                1. Else:
-                    1. Let |outputTensor| be the output tensor returned by |this|.{{MLGraph/[[implementation]]}}.
-                    1. If the kind of |value| is not compatible with the value type of |outputTensor|, then throw a {{DataError}} {{DOMException}} and stop.
-                    1. Let |outputSize| be 1.
-                    1. For each |dimension| of dimensions of |outputTensor|:
-                        1. Set |outputSize| to the product of |outputSize| and |dimension|.
-                    1. If |outputSize| is greater than the length of |value|, then:
-                        1. Throw a {{DataError}} {{DOMException}} and stop.
-                    1. Else:
-                        1. Set the values of |value| to the values of |outputTensor|.
-            1. Return {{undefined}}.
-
-            Issue: Describe the algorithm steps for |this|.{{MLGraph/[[context]]}} created from {{WebGLRenderingContext}} and {{GPUDevice}}.
-        </div>
+        The context of type {{MLContext}} associated with this {{MLCommandEncoder}}.
+
+    : <dfn>\[[implementation]]</dfn>
+    ::
+        The underlying implementation provided by the User Agent.
 </dl>
 
-### Examples ### {#compilation-examples}
+### Weight Preprocessing ### {#api-mlcommandencoder-graph-preprocessing}
+Record the initialization of the {{MLGraph}} with constant inputs {{MLNamedGPUInputs}} such as weight tensors. This is a necessary step for optimal performance as it gives the platform an opportunity to prepare and optimize constant input data for the subsequent execution of the graph. This method should only be called once per graph.
 
-<div class="example">
-The following code showcases the computation with dynamic input dimensions.
-<pre highlight="js">
-function sizeOfShape(array) {
-  return array.reduce(
-      (accumulator, currentValue) => accumulator * currentValue);
-}
+<script type=idl>
+partial interface MLCommandEncoder {
+  undefined initializeGraph(MLGraph graph, MLNamedGPUInputs constants);
```

Comment:
Does it mean the caller should define the weight operands by `MLGraphBuilder::input` when building the graph?

This may lead to different ways of defining weights:
1.  Using `MLGraphBuilder::constant` for `MLContext` created from `MLContextOptions`
2.  Using `MLGraphBuidler::input` for `MLContext` created from `GPUDevice`

Should we also define `MLContext::initializeGraph` (and probably also with `MLContext::initializeGraphAsync`) that takes `MLNamedArrayInputs`?

Then the `MLGraphBuilder::constant` could just create a named placeholder (or just overload `MLGraphBuilder::input`) that will be bound to different buffer type according to the context type when `initializeGraph` invoked.

Reply:
Or we can just remove the second param to `initializeGraph`. Your comment makes me realize that the second param is entirely redundant to what already in the graph. There is already an overload for `constant` in the graph builder interface that would let the caller pass in a different type of resource e.g. `GPUBuffer`. 

I don't think we need `initializeGraphAsync`. The process of recording the workload to the command buffer typically happens very quickly and entirely on the CPU side. It's hardly a potentially block call and shouldn't have the need to offload it to an async call.

Reply:
> The process of recording the workload to the command buffer typically happens very quickly and entirely on the CPU side. It's hardly a potentially block call and shouldn't have the need to offload it to an async call.

This is true for `MLCommandEncoder`.

However, we may need to define how the "weight preprocessing" happens in a default `MLContext`. In current design, I suppose it would happen upon `MLGraphBuilder::build` being invoked. If the device type is `gpu`, the steps of weight processing might be:
1. record the initialization of the graph
2. submit the command buffer to the queue
3. wait for queue finishes the initialization

As the `build` is a sync method, it may block the calling thread while waiting for GPU queue execution.

> There is already an overload for `constant` in the graph builder interface that would let the caller pass in a different type of resource e.g. `GPUBuffer`.

I think the constant / weight uploading and preprocessing should also be context-based. Correspondingly the overload for `constant` in the graph builder could be removed. Something like:

| Context type | Constant / weight uploading and preprocessing method | Buffer type |
|---|---|---|
| "default" | `MLContext::initializeGraphAsync(graph, constants)` and <br> `MLContext::initializeGraph(graph, constants)` (only for worker) | `ArrayBufferView` |
| "webgpu" | `MLCommandEncoder::initializeGraph(graph, constants)` | `GPUBuffer` |

Hopefully, this would also make the graph builder implementation simpler, because it won't need to hold constant buffers any more.

Reply:
Weight preprocessing in the "default" context case can be done on the first call to `compute`. I don't think there is a need for a dedicated `initializeGraph` call there. 

The `MLCommandEncoder` case needs an `initializeGraph` call because the purpose of the call isn't to execute the graph, but only to record the workload and return it in a command buffer. By producing the initialize workload in a command buffer, it allows more flexibility for the caller to choose how/when to submit them for execution.

Reply:
Please take a look at the latest commit 59bbbe9a4821c6e9b6b5c11384bf9c6e50c60533.

Reply:
Thanks @wchao1115 , the latest commit looks good.

> Weight preprocessing in the "default" context case can be done implicitly on the first call to `compute`.

It's possible. At minimum, we should define when and where this happens. Otherwise, this would lead to some delay in the first call to `compute` that the caller may not expect. IMO, the flexibility and explicit control of the graph initialization by `MLCommandEncoder::initializeGraph` is useful for a good user experience, because the caller can ensure the graph is ready to compute in an efficient way. We may think about supporting it in the "default" context.

Given this is out of scope of this PR focusing on graph execution, I propose we deal with it with a separate issue. WDYT?

Reply:
For the default context, I think your suggestion to have weight processing done as the last step of `MLGraphBuilder.build` makes a lot of sense to me, given that all the constant operands are already in place in the graph builder state at that time. I can put a note section to clarify that if it's useful. 

I still think having a separate `initializeGraph` call for the WebGPU context case is useful since it'll allow the app to fold the initialize workload to other graphics workload that they already have in the same queue submission.

Reply:
> For the default context, I think your suggestion to have weight processing done as the last step of `MLGraphBuilder.build` makes a lot of sense to me, given that all the constant operands are already in place in the graph builder state at that time. I can put a note section to clarify that if it's useful.

I think it's useful. But there is still an issue that the sync `build` method may block the calling thread until the graph initialization is done. This may be even worse when the device type is gpu that involves uploading weights to gpu.

> I still think having a separate `initializeGraph` call for the WebGPU context case is useful since it'll allow the app to fold the initialize workload to other graphics workload that they already have in the same queue submission.

I'd agree. Then we should also note that for the webgpu context, the `MLGraphBuilder.build` won't process the weights. And this would make the weights processing a bit complex because the `MLGraph` implementation may need to hold the unprocessed weights until `initializeGraph` being called. Actually I like the `initializeGraph` that takes the constant buffers. It allows the implementation simply pass the caller-supplied constant buffers to native ML API without holding a copy or reference.

Reply:
> I can put a note section to clarify that if it's useful.

Added in this commit 0ed918fac4da4ada3501fffe2513da429e8e1a11 . Please take a look.

Reply:
Track the async build issue in https://github.com/webmachinelearning/webnn/issues/263.